Fix the magic bytes handling (again, possibly)

whot opened the pull request:
This supersedes #6 and hopefully works for that device too.

Turns out rusb's `read_string_descriptor()` **cannot** work on some devices because it expects an even-length buffer and the Kamvas 12 returns a buffer length 19. So we need to do this manually. And to be doubly safe, do exactly what hid-uclogic does with two "normal" string reads and two "special" reads that return the buffer as-is, including header bytes.

The PR could be rebased but that's too much effort at this point...

cc @bentiss

whot committed:
Move handling the string descriptor into a helper function

whot committed:
Use find() instead of filter for our single match

There's only one language with that particular language ID, so let's use
find() + map() instead of filter() + for_each().

whot committed:
Bubble errors up if we fail to read a string

bentiss committed:
magic_bytes: do not mess up with the encoding

rusb does a `String::from_utf16`:
https://docs.rs/rusb/latest/src/rusb/device_handle.rs.html#757

However, this converts any special characters into utf16, and when we take
the bytes out of them, we get something different than the USB output.

Convert back the data to utf16, and then into big endian to restore the
original USB data.

bentiss commented on src/main.rs:
```diff
@@ -13,51 +13,67 @@ fn string_descriptor(
+    // uclogic_params_pen_init_v2()
+    match string_descriptor(&handle, &lang, 100) {
+        Ok(s) => {
+            if s.as_bytes().len() == 12 {
```
This should be `if bytes.len() >= 12`, my H610 returns an array of 14 bytes.

whot replied:
fixed by your commit now

bentiss commented:
I still got the "udev ignores everything but the first properties" bug. I wonder if we should not manually convert the fw id from UTF8 ourself instead of relying on rusb's implementation.

bentiss committed:
Do not require an exact 12 length magic bytes string for v1

Turns out the kernel requests only a 12 bytes length string, hence 12 is
a good number to check against in the kernel. Here we read more than
that so let's make this the lower threshold.

bentiss committed:
firmware_id: make this compatible with udev

Turns out my firmware id has an extra `\0` at the end, and when udev
reads this it considers the end of the stream.

Strip it and also quotes the string so weird characters gets escaped.

bentiss commented on src/main.rs:
```diff
@@ -1,21 +1,46 @@
+        return Err(rusb::Error::BadDescriptor);
+    }
+
+    Ok(buf[..len].to_vec())
```
I won't argue too much here, but I'm not sure there is a point in returning the first 2 bytes. They are USB implementation detail, and the payload is actually `2..len`.

But OTOH, this allows to check from udev that the string is complete, so meh?

whot replied:
there's no point since they're effectively fixed but this is solely to match the data buffer with the one the kernel driver uses so we have the same offsets/indices. which I think might be useful long-term.

whot commented:
ok, immediate bugs fixed, let's merge this! thanks

whot committed:
Match the kernel's sequence of usb queries

In the kernel EPIPE doesn't end the sequence of usb queries, it
continues. Let's match that.

And like the kernel require that the 100 query returns a 12-byte string.

whot committed:
Read the string descriptor for the HUION_MAGIC_BYTES manually

rusb's read_string_descriptor() enforces UTF16 with a peculiar check:

https://docs.rs/rusb/latest/src/rusb/device_handle.rs.html#757
    if len < 2 || buf[0] != len as u8 || len & 0x01 != 0 {
                    return Err(Error::BadDescriptor);
    }

This makes sense since it expects UTF16 but the Kamvas 12 returns a
buffer length is 19 - one byte for the length, one byte by the device
type (0x3) followed by 17 data bytes.  But we never get past this check
because rusb expects an even buffer length.
On the Huion Inspiroy 2S the total buffer is 20, so it works there.

So split this up: for the two queries that result in the magic bytes
being printed (index 200 and 100) use a custom message that returns the
full buffer including the two header bytes - this matches the
hid-uclogic driver which does all the checking on bytes 2+.

For the firmware and the pad mode use a normal string descriptor read,
just like hid-uclogic does.
